keps/sig-security/2763-ambient-capabilities/README.md

```diff
@@ -0,0 +1,929 @@
+
+  **Pros and Cons**
+
+  :+1: Simple add and drop API is easy to use from a user perspective.
+
+  :+1: No new field needs to added to the K8S API Capabilities object.
+
+  :-1: How would we tell the difference at API validation time? I assume we'd have to look at mustRunAsNonRoot, because the apiserver has no idea whether a textual runAsUser is root or not. mustRunAsNonRoot it designed to fail at the last minute, in the Kubelet, which is bad UX. This makes this approach very awkward. Different behavior for root vs non-root might be confusing for users.
+
+  :-1: Since we are changing the default behavior for non-root containers we might break existing user containers. Issues with linux capabilities have proven themselves to be very hard to debug in the past.
+
+- **Option 2:** Add new field to [Capabilities](https://pkg.go.dev/k8s.io/api/core/v1#Capabilities)
+
+  When a capability is added using this field only then does it get added to the ambient set in addition to inheritable, permitted, bounding and effective sets. The behavior is consistent for root and non-root containers.
+
+  ```diff
+  type Capabilities struct {
+    // Added capabilities
+    // +optional
+    Add []Capability `json:"add,omitempty" protobuf:"bytes,1,rep,name=add,casttype=Capability"`
+    // Removed capabilities
+    // +optional
+    Drop []Capability `json:"drop,omitempty" protobuf:"bytes,2,rep,name=drop,casttype=Capability"`
+  + // Ambient capabilities to add
+  + // +optional
+  + Ambient []Capability `json:"ambient,omitempty" protobuf:"bytes,3,rep,name=ambient,casttype=Capability"`
+  }
+  ```
+
+  **Pros and Cons**
+
+  :+1: Since this is a new field existing user workloads are unaffected by this change.
+
+  :+1: The capabilities struct only really translates to Linux Capabilities, adding a new field does make it easier for users who understand Linux capabilities to configure them.
+
+  :+1: Ambient caps are useful for root containers as well. A root process can control what caps it gives to child processes that are running as non-root. The capabilities list list for these non-root processes may not be the same as the root caps in Add.
+
+  :-1: Requires changes to the K8S API Capabilities object.
+
+<<[/UNRESOLVED]>>
+
+`Ambient` capabilities will adhere to the following rules:
+
+1. Ambient capabilities can only be added.
+2. Default capabilities are not added to the ambient capabilities set. By default the container has an empty ambient capability set,
+3. Only capability that are explicitly added in the manifest will be added to the ambient set.
+4. Since the ambient capability set obeys the invariant that no capability can ever be ambient if it is not both permitted and inheritable, adding a capability to the ambient set will also add it to the permitted and inheritable set.
+
+#### Restricted ambient capabilities.
+
+<<[UNRESOLVED what is the set of capabilities that we should allow to be made ambient]>>
```

Comment:
Previously when this was discussed PSP's were deprecated and there wasn't alternative. But we now have a PodSecurity admission controller in beta and enabled by default as of 1.23.

So I think if there are any restrictions it should be done in the PodSecurity admission controller. Most should probably be under the 'Privileged' category.

Reply:
@Jc2k are you thinking of a change to the [pod security standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/)? These are what you can enforce with Pod Security admission.

Reply:
It depends.

For context, let me stress again that when this section of the KEP were first discussed, the admission controller wasn't in beta (if it existed at all) so it wasn't considered a valid solution to this `UNRESOLVED` part of the KEP.

In the case where we re-use `securityContext.capabilities.add`, the existing standards look like they are valid for root and non-root containers. Neither would be allowed to use `CAP_SYS_ADMIN` or `CAP_DAC_OVERRIDE` etc unless they were using the `privileged` policy level. This is good.

If we are **not** reusing `securityContext.capabilities.add`, then I think the standards should be updated anyway. They currently cover all the ways you can add capabilities to containers. If there is a new way to add a capabitility to a container, it should be covered by the standards.